File: sdk/mobile/android/src/main/kotlin/com/summit/odps/android/TelemetryModels.kt
package com.summit.odps.android

import java.util.UUID

/**
 * Describes a telemetry event collected by the ODPS pipeline.
 */
data class TelemetryEvent(
  val eventId: String = UUID.randomUUID().toString(),
  val type: String,
  val payload: Map<String, Any?> = emptyMap(),
  val metrics: List<Metric> = emptyList()
)

/**
 * Supported metric aggregation types that can be protected via local differential privacy.
 */
enum class MetricType {
  COUNT,
  SUM
}

/**
 * Numeric metric emitted from the device prior to local-DP sanitisation.
 */
data class Metric(
  val name: String,
  val type: MetricType,
  val value: Double
)

File: sdk/mobile/ios/Sources/ODPS/TelemetryModels.swift
import Foundation

public struct TelemetryEvent: Equatable {
    public let eventId: String
    public let type: String
    public let payload: [String: AnyHashable]
    public let metrics: [Metric]

    public init(eventId: String = UUID().uuidString,
                type: String,
                payload: [String: AnyHashable] = [:],
                metrics: [Metric] = []) {
        self.eventId = eventId
        self.type = type
        self.payload = payload
        self.metrics = metrics
    }
}

public enum MetricType {
    case count
    case sum
}

public struct Metric: Equatable {
    public let name: String
    public let type: MetricType
    public let value: Double

    public init(name: String, type: MetricType, value: Double) {
        self.name = name
        self.type = type
        self.value = value
    }
}
